File: 4 Algorithms on Strings/Week 2 Burrows-Wheeler Transform and Suffix Arrays/Programming-Assignment-2/bwtinverse/bwtinverse.cpp
#include <algorithm>
#include <iostream>
#include <string>
#include <vector>
#include <utility>

using std::cin;
using std::cout;
using std::endl;
using std::string;
using std::vector;
using std::pair;
using std::make_pair;

string InverseBWT(const string& bwt) {
  string text = "";

  // write your code here
  vector<pair<char, int>> first(bwt.size());
  for (int i = 0; i < bwt.size(); i++){
      first[i] = make_pair(bwt[i], i);
  }
  sort(first.begin(), first.end());
  
  pair<char, int> elem = first[0];
  for (int i = 0; i < bwt.size(); i++) {
  	elem = first[elem.second];
  	text += elem.first;
  }

  return text;
}

int main() {
  string bwt;
  cin >> bwt;
  cout << InverseBWT(bwt) << endl;
  return 0;
}

File: 4 Algorithms on Strings/Week 2 Burrows-Wheeler Transform and Suffix Arrays/Programming-Assignment-2/bwtinverse/bwtinverse.py
# python3
import sys

def InverseBWT(bwt):
    # write your code here
    result = ""
    n = len(bwt)
    last = [(char, index) for (index, char) in enumerate(bwt)]
    first = sorted(last)
    
    elem = first[0]
    for i in range(n):
        elem = first[elem[1]]
        result += elem[0]
    
    return result


if __name__ == '__main__':
    bwt = sys.stdin.readline().strip()
    print(InverseBWT(bwt))